// @hint/hint-webpack-config/dist/tests/is-installed.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const path = require("path");
const mock = require("mock-require");
const utils_1 = require("@hint/utils");
const utils_tests_helpers_1 = require("@hint/utils-tests-helpers");
const { getHintPath } = utils_1.test;
const { loadJSONFile } = utils_1.fs;
const webpackDestPath = path.join(__dirname, 'fixtures', 'valid', 'package.json');
const webpackConfig = loadJSONFile(webpackDestPath);
const hintPath = getHintPath(__filename, true);
const tests = [
    {
        before() {
            const loadPackage = () => {
                return webpackConfig;
            };
            mock('@hint/utils/dist/src/packages/load-package', { loadPackage });
        },
        name: 'If valid configuration file exists and webpack is installed should pass',
        path: path.join(__dirname, 'fixtures', 'valid')
    },
    {
        before() {
            const loadPackage = () => {
                throw new Error('error');
            };
            mock('@hint/utils/dist/src/packages/load-package', { loadPackage });
        },
        name: 'If valid configuration file exists but webpack is not installed should fail',
        path: path.join(__dirname, 'fixtures', 'valid'),
        reports: [{ message: `webpack is not installed in your project.` }]
    }
];
utils_tests_helpers_1.testLocalHint(hintPath, tests, {
    parsers: ['webpack-config'],
    serial: true
});
