// @hint/parser-manifest/dist/src/parser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const types_1 = require("hint/dist/src/lib/types");
const normalize_string_1 = require("@hint/utils/dist/src/misc/normalize-string");
const is_http_1 = require("@hint/utils/dist/src/network/is-http");
const is_https_1 = require("@hint/utils/dist/src/network/is-https");
const json_parser_1 = require("@hint/utils/dist/src/json-parser");
const schema_validator_1 = require("@hint/utils/dist/src/schema-validation/schema-validator");
const schema = require('./schema.json');
class ManifestParser extends types_1.Parser {
    constructor(engine) {
        super(engine, 'manifest');
        this.fetchEndEventName = 'fetch::end::manifest';
        this.fetchErrorEventName = 'fetch::error::manifest';
        this.fetchStartEventName = 'fetch::start::manifest';
        this.parseEndEventName = 'parse::end::manifest';
        this.parseErrorSchemaEventName = 'parse::error::manifest::schema';
        this.parseJSONErrorEventName = 'parse::error::manifest::json';
        this.fetchedManifests = new Set();
        engine.on('element::link', this.fetchManifest.bind(this));
        engine.on('fetch::end::manifest', this.validateManifest.bind(this));
        engine.on('scan::end', this.onScanEnd.bind(this));
    }
    onScanEnd() {
        this.fetchedManifests.clear();
    }
    async fetchManifest(elementFound) {
        const { element, resource } = elementFound;
        if (!is_http_1.isHTTP(resource) && !is_https_1.isHTTPS(resource)) {
            return;
        }
        if (normalize_string_1.normalizeString(element.getAttribute('rel')) !== 'manifest') {
            return;
        }
        const hrefValue = normalize_string_1.normalizeString(element.getAttribute('href'));
        if (!hrefValue) {
            return;
        }
        const manifestURL = element.resolveUrl(hrefValue);
        if (this.fetchedManifests.has(manifestURL)) {
            return;
        }
        await this.engine.emitAsync(this.fetchStartEventName, { resource });
        let manifestNetworkData;
        try {
            manifestNetworkData = await this.engine.fetchContent(manifestURL);
        }
        catch (error) {
            await this.engine.emitAsync(this.fetchErrorEventName, {
                element,
                error,
                hops: [manifestURL],
                resource: manifestURL
            });
            return;
        }
        if (this.fetchedManifests.has(manifestURL)) {
            return;
        }
        await this.engine.emitAsync(this.fetchEndEventName, {
            element,
            request: manifestNetworkData.request,
            resource: manifestURL,
            response: manifestNetworkData.response
        });
    }
    async validateManifest(fetchEnd) {
        const { resource, response } = fetchEnd;
        if (this.fetchedManifests.has(resource)) {
            return;
        }
        this.fetchedManifests.add(resource);
        if (response.statusCode >= 400) {
            return;
        }
        await this.engine.emitAsync(`parse::start::manifest`, { resource });
        let result;
        try {
            result = json_parser_1.parseJSON(response.body.content);
        }
        catch (e) {
            await this.engine.emitAsync(this.parseJSONErrorEventName, {
                error: e,
                resource
            });
            return;
        }
        const validationResult = schema_validator_1.validate(schema, result.data, result.getLocation);
        if (!validationResult.valid) {
            await this.engine.emitAsync(this.parseErrorSchemaEventName, {
                error: new Error('Invalid manifest'),
                errors: validationResult.errors,
                groupedErrors: validationResult.groupedErrors,
                prettifiedErrors: validationResult.prettifiedErrors,
                resource
            });
            return;
        }
        await this.engine.emitAsync(this.parseEndEventName, {
            getLocation: result.getLocation,
            parsedContent: validationResult.data,
            resource
        });
    }
}
exports.default = ManifestParser;
