// hint/dist/src/lib/hint-context.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const category_1 = require("@hint/utils/dist/src/types/category");
const get_html_code_snippet_1 = require("@hint/utils/dist/src/report/get-html-code-snippet");
class HintContext {
    constructor(hintId, engine, severity, options, meta, ignoredUrls) {
        this.id = hintId;
        this.options = options;
        this.meta = meta;
        this.engine = engine;
        this.severity = severity;
        this.ignoredUrls = ignoredUrls;
        Object.freeze(this);
    }
    get engineKey() {
        return this.engine;
    }
    get pageDOM() {
        return this.engine.pageDOM;
    }
    get pageContent() {
        return this.engine.pageContent;
    }
    get pageHeaders() {
        return this.engine.pageHeaders;
    }
    get targetedBrowsers() {
        return this.engine.targetedBrowsers;
    }
    get hintOptions() {
        if (Array.isArray(this.options)) {
            return this.options[1];
        }
        return null;
    }
    evaluate(source) {
        return this.engine.evaluate(source);
    }
    fetchContent(target, headers) {
        return this.engine.fetchContent(target, headers);
    }
    querySelectorAll(selector) {
        return this.engine.querySelectorAll(selector);
    }
    findProblemLocation(element, offset) {
        if (offset) {
            return element.getContentLocation(offset);
        }
        return element.getLocation();
    }
    report(resource, message, options = {}) {
        const { codeSnippet, element, severity } = options;
        let sourceCode = null;
        let position = options.location || null;
        if (element) {
            position = this.findProblemLocation(element, position);
            sourceCode = get_html_code_snippet_1.getHTMLCodeSnippet(element);
        }
        this.engine.report({
            category: (this.meta && this.meta.docs && this.meta.docs.category) ? this.meta.docs.category : category_1.Category.other,
            codeLanguage: options.codeLanguage,
            hintId: this.id,
            location: position || { column: -1, line: -1 },
            message,
            resource,
            severity: severity || this.severity,
            sourceCode: codeSnippet || sourceCode || ''
        });
    }
    on(event, listener) {
        this.engine.onHintEvent(this.id, event, listener);
    }
    isUrlIgnored(resource) {
        return this.ignoredUrls.some((urlIgnored) => {
            return urlIgnored.test(resource);
        });
    }
    get language() {
        return this.engine.language;
    }
}
exports.HintContext = HintContext;
