// @hint/hint-axe/dist/src/meta/language.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const category_1 = require("hint/dist/src/lib/enums/category");
const hint_scope_1 = require("hint/dist/src/lib/enums/hint-scope");
const i18n_import_1 = require("../i18n.import");
const meta = {
    docs: {
        category: category_1.Category.accessibility,
        description: i18n_import_1.getMessage('language_description', 'en'),
        name: i18n_import_1.getMessage('language_name', 'en')
    },
    getDescription(language) {
        return i18n_import_1.getMessage('language_description', language);
    },
    getName(language) {
        return i18n_import_1.getMessage('language_name', language);
    },
    id: 'axe/language',
    schema: [{
            additionalProperties: false,
            properties: {
                'html-has-lang': { enum: ['off', 'warning', 'error'], type: 'string' },
                'html-lang-valid': { enum: ['off', 'warning', 'error'], type: 'string' },
                'html-xml-lang-mismatch': { enum: ['off', 'warning', 'error'], type: 'string' },
                'valid-lang': { enum: ['off', 'warning', 'error'], type: 'string' }
            }
        }],
    scope: hint_scope_1.HintScope.any
};
exports.default = meta;
