// @hint/hint-axe/dist/src/meta/name-role-value.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const category_1 = require("hint/dist/src/lib/enums/category");
const hint_scope_1 = require("hint/dist/src/lib/enums/hint-scope");
const i18n_import_1 = require("../i18n.import");
const meta = {
    docs: {
        category: category_1.Category.accessibility,
        description: i18n_import_1.getMessage('nameRoleValue_description', 'en'),
        name: i18n_import_1.getMessage('nameRoleValue_name', 'en')
    },
    getDescription(language) {
        return i18n_import_1.getMessage('nameRoleValue_description', language);
    },
    getName(language) {
        return i18n_import_1.getMessage('nameRoleValue_name', language);
    },
    id: 'axe/name-role-value',
    schema: [{
            additionalProperties: false,
            properties: {
                'aria-hidden-focus': { enum: ['off', 'warning', 'error'], type: 'string' },
                'button-name': { enum: ['off', 'warning', 'error'], type: 'string' },
                'empty-heading': { enum: ['off', 'warning', 'error'], type: 'string' },
                'input-button-name': { enum: ['off', 'warning', 'error'], type: 'string' },
                'link-name': { enum: ['off', 'warning', 'error'], type: 'string' }
            }
        }],
    scope: hint_scope_1.HintScope.any
};
exports.default = meta;
