// @hint/hint-axe/dist/src/meta/semantics.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const category_1 = require("hint/dist/src/lib/enums/category");
const hint_scope_1 = require("hint/dist/src/lib/enums/hint-scope");
const i18n_import_1 = require("../i18n.import");
const meta = {
    docs: {
        category: category_1.Category.accessibility,
        description: i18n_import_1.getMessage('semantics_description', 'en'),
        name: i18n_import_1.getMessage('semantics_name', 'en')
    },
    getDescription(language) {
        return i18n_import_1.getMessage('semantics_description', language);
    },
    getName(language) {
        return i18n_import_1.getMessage('semantics_name', language);
    },
    id: 'axe/semantics',
    schema: [{
            additionalProperties: false,
            properties: {
                'heading-order': { enum: ['off', 'warning', 'error'], type: 'string' },
                'landmark-banner-is-top-level': { enum: ['off', 'warning', 'error'], type: 'string' },
                'landmark-complementary-is-top-level': { enum: ['off', 'warning', 'error'], type: 'string' },
                'landmark-contentinfo-is-top-level': { enum: ['off', 'warning', 'error'], type: 'string' },
                'landmark-main-is-top-level': { enum: ['off', 'warning', 'error'], type: 'string' },
                'landmark-no-duplicate-banner': { enum: ['off', 'warning', 'error'], type: 'string' },
                'landmark-no-duplicate-contentinfo': { enum: ['off', 'warning', 'error'], type: 'string' },
                'landmark-one-main': { enum: ['off', 'warning', 'error'], type: 'string' },
                'landmark-unique': { enum: ['off', 'warning', 'error'], type: 'string' },
                'layout-table': { enum: ['off', 'warning', 'error'], type: 'string' },
                'p-as-heading': { enum: ['off', 'warning', 'error'], type: 'string' },
                'page-has-heading-one': { enum: ['off', 'warning', 'error'], type: 'string' }
            }
        }],
    scope: hint_scope_1.HintScope.any
};
exports.default = meta;
