// @hint/hint-axe/dist/src/meta/time-and-media.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const category_1 = require("hint/dist/src/lib/enums/category");
const hint_scope_1 = require("hint/dist/src/lib/enums/hint-scope");
const i18n_import_1 = require("../i18n.import");
const meta = {
    docs: {
        category: category_1.Category.accessibility,
        description: i18n_import_1.getMessage('timeAndMedia_description', 'en'),
        name: i18n_import_1.getMessage('timeAndMedia_name', 'en')
    },
    getDescription(language) {
        return i18n_import_1.getMessage('timeAndMedia_description', language);
    },
    getName(language) {
        return i18n_import_1.getMessage('timeAndMedia_name', language);
    },
    id: 'axe/time-and-media',
    schema: [{
            additionalProperties: false,
            properties: {
                'audio-caption': { enum: ['off', 'warning', 'error'], type: 'string' },
                blink: { enum: ['off', 'warning', 'error'], type: 'string' },
                'meta-refresh': { enum: ['off', 'warning', 'error'], type: 'string' }
            }
        }],
    scope: hint_scope_1.HintScope.any
};
exports.default = meta;
